src/app-router.js

```diff
@@ -40,53 +34,37 @@ export class AppRouter extends Router {
     });
   }
 
-  dequeueInstruction() {
-    if (this.isNavigating) {
-      return;
-    }
-
-    var instruction = this.queue.shift();
-    this.queue = [];
-
-    if (!instruction) {
-      return;
-    }
-
-    this.isNavigating = true;
-    this.events.publish('router:navigation:processing', instruction);
-
-    var context = this.createNavigationContext(instruction);
-    var pipeline = this.pipelineProvider.createPipeline(context);
+  dequeueInstruction(instructionCount = 0) {
+    return Promise.resolve().then(() => {
+      if (this.isNavigating && !instructionCount) {
+        return;
+      }
 
-    pipeline.run(context).then(result => {
-      this.isNavigating = false;
+      let instruction = this.queue.shift();
+      this.queue = [];
 
-      if (!(result && 'completed' in result && 'output' in result)) {
-        throw new Error(`Expected router pipeline to return a navigation result, but got [${JSON.stringify(result)}] instead.`);
+      if (!instruction) {
+        return;
       }
 
-      if (result.completed) {
-        this.history.previousFragment = instruction.fragment;
-      }
+      this.isNavigating = true;
 
-      if (result.output instanceof Error) {
-        console.error(result.output);
-        this.events.publish('router:navigation:error', { instruction, result });
+      if (!instructionCount) {
+        this.events.publish('router:navigation:processing', { instruction });
+      } else if (instructionCount > this.maxInstructionCount) {
```

Comment:
We could try to restore the last known good fragment instead of just giving up, but this helps avoid the more immediately problem of getting stuck in an infinite redirect loop.